include/field.hxx

```diff
@@ -251,6 +251,27 @@ inline T filledFrom(const T& f, BoutReal fill_value) {
   return result;
 }
 
+/// Return a field of some type derived from Field, with metadata copied from
+/// another field and a data array allocated and filled using a callable e.g. lambda function
+///
+/// e.g.
+///   Field3D result = filledFrom(field, [&](const auto& index) {
+///                                          return ...;
+///                                      });
+/// 
+/// An optional third argument is the region string
+template <
+    typename T, typename Function,
+    typename = decltype(std::declval<Function&>()(std::declval<typename T::ind_type&>()))>
```

Comment:
Ahh, C++! :smile: The only bit I'm not sure of is the reference in `declval<Function&>` -- what am I missing?

Reply:
I'm also a bit unclear on this magic incantation. It's adapted from https://stackoverflow.com/questions/50859115/enable-if-template-param-is-lambda-with-particular-signature

Reply:
Oh, I see, it's to guard against strange types! If `F` had a call operator like `void operator(ind_type) &&`, which is _rvalue_ qualified, it can only be called on a temporary. But we definitely have an _lvalue_ here, that is, a named variable. `F&` is a mutable reference, so it must be an _lvalue_, so it ensures that this will definitely be the right `operator()` overload.

Phew!

This might be a FIXME for when we get to move to C++17 and it can be `std::is_invocable`